middle_end/flambda2.0/from_lambda/lambda_to_flambda_primitives.ml

```diff
@@ -174,6 +174,81 @@ let bigarray_unbox_value_to_store kind =
       "Don't know how to unbox a fabricated expression to \
        store it in a bigarray"
 
+let bigarray_dim_bound b dimension =
+  H.Prim (Unary (Bigarray_length { dimension }, b))
+
+let bigarray_check_bound idx bound =
+  H.Binary (Int_comp (I.Naked_immediate, Unsigned, Lt), idx, bound)
+
+(* CR Gbury: this function in effect duplicates the bigarray_lenght access:
+             one is done in the validity check, and one in the final offset
+             computation, whereas cmmgen let-binds this access. It might
+             matter for the performance, although the processor cache might
+             make it not matter at all. *)
+let bigarray_indexing layout b args =
+  let num_dim = List.length args in
+  let rec aux dim delta_dim = function
+    | [] -> assert false
+    | [idx] ->
+      let bound = bigarray_dim_bound b dim in
+      let idxn = untag_int idx in
+      let check = bigarray_check_bound idxn bound in
+      [check], idx
+    | idx :: r ->
+      let checks, rem = aux (dim + delta_dim) delta_dim r in
+      let bound = bigarray_dim_bound b dim in
+      let idxn = untag_int idx in
+      let check = bigarray_check_bound idxn bound in
+      (* CR gbury: because we tag bound, and the tagged multiplication
+         untags it, we might be left with a needless zero-extend here. *)
+      let tmp =
+        H.Prim (Binary (Int_arith(I.Tagged_immediate, Mul), rem,
+                        Prim (Unary (Box_number Untagged_immediate, bound))))
+      in
+      let offset =
+        H.Prim (Binary (Int_arith(I.Tagged_immediate, Add), tmp, idx))
+      in
+      check :: checks, offset
+  in
+  match layout with
+  | P.C ->
```

Comment:
Here and also below, the general convention in the Flambda code base is to type-annotate either the let-bound variable or the function parameter, etc.  This should overall reduce the clutter of module name qualifications.